First, we import some of the packages that we will use. Packages are code written by someone else that we can use. In this example, we are using the entire "pandas" package and the package is reference by the variable "pd". In addition, we are using the "variance_inflation_factor" method from the "statsmodel"s package.

In [1]:
import pandas as pd
from statsmodels.stats.outliers_influence import variance_inflation_factor

## 1. Importing Data
#### 1.1 Importing data from excel to dataframe
This code chunk imports the sheet "T18" from the excel "t-18-22.xls" and store it in the "type_of_dwelling_df" variable.

In [2]:
type_of_dwelling_df = pd.read_excel('t18-22.xls', 'T18')

## 2. Basic Data Exploration
#### 2.1 First/last 5 records
This code chunk the first 5 records. We can also show the last 5 records using "type_of_dwelling_df.tail()".

In [3]:
type_of_dwelling_df.head() 

Planning Area  1- and 2-Room Flats  3-Room Flats  4-Room Flats  \
0    Ang Mo Kio                 5122         26090         13993   
1         Bedok                 4233         23792         21673   
2        Bishan                  444          2371          9920   
3   Bukit Batok                   27         10970         15245   
4   Bukit Merah                11319         16773         14787   

   5-Room and Executive Flats  Condominiums and Other Apartments  \
0                        6271                               3661   
1                       14003                              15148   
2                        7510                               3570   
3                        8172                               8447   
4                        9534                               2709   

   Landed Properties  Others  
0               4177     391  
1              11765     611  
2               3162      53  
3               1164     108  
4                160     344

#### 2.2 Columns

In [4]:
type_of_dwelling_df.columns

Index(['Planning Area', '1- and 2-Room Flats', '3-Room Flats', '4-Room Flats',
       '5-Room and Executive Flats', 'Condominiums and Other Apartments',
       'Landed Properties', 'Others'],
      dtype='object')

#### 2.3 Basic column statistics

In [5]:
type_of_dwelling_df.describe()

1- and 2-Room Flats  3-Room Flats  4-Room Flats  \
count            36.000000     36.000000     36.000000   
mean           1452.055556   6381.055556  10150.666667   
std            2493.521353   7286.758852   9735.563114   
min               0.000000      0.000000      0.000000   
25%               0.000000    120.750000    763.750000   
50%             281.000000   2972.500000   8058.000000   
75%            1422.750000  11955.500000  15047.750000   
max           11319.000000  26090.000000  30448.000000   

       5-Room and Executive Flats  Condominiums and Other Apartments  \
count                   36.000000                          36.000000   
mean                  8148.194444                        3666.194444   
std                   8620.261944                        3059.439284   
min                      0.000000                           0.000000   
25%                    178.750000                        1452.500000   
50%                   6386.500000                        3615.500000   
75%                  12469.000000                        4715.500000   
max                  29644.000000                       15148.000000   

       Landed Properties      Others  
count            36.0000   36.000000  
mean           1803.0000  148.027778  
std            2732.5237  169.407284  
min               0.0000    0.000000  
25%             172.0000   16.750000  
50%             797.5000  100.000000  
75%            1986.5000  187.250000  
max           11765.0000  611.000000

## 3. Using dataframe
### 3.1 Changing column name
Column name should preferably start with alphabet and not contain space

In [6]:
type_of_dwelling_df.columns = ['planning_area', 'flats_1_2_room', 'flats_3_room', 'flats_4_room',
       'flats_5_room_and_executive', 'condominiums_and_other_apartments',
       'landed_properties', 'others']
type_of_dwelling_df.head()

planning_area  flats_1_2_room  flats_3_room  flats_4_room  \
0    Ang Mo Kio            5122         26090         13993   
1         Bedok            4233         23792         21673   
2        Bishan             444          2371          9920   
3   Bukit Batok              27         10970         15245   
4   Bukit Merah           11319         16773         14787   

   flats_5_room_and_executive  condominiums_and_other_apartments  \
0                        6271                               3661   
1                       14003                              15148   
2                        7510                               3570   
3                        8172                               8447   
4                        9534                               2709   

   landed_properties  others  
0               4177     391  
1              11765     611  
2               3162      53  
3               1164     108  
4                160     344

#### 3.2 Getting subset of columns
We can get column by using a dot followed by the column name. If we did not rename the columns, this way of access column will not work for those columns with space and those that starts with a number.

In [7]:
type_of_dwelling_df.planning_area

0          Ang Mo Kio
1               Bedok
2              Bishan
3         Bukit Batok
4         Bukit Merah
5       Bukit Panjang
6         Bukit Timah
7              Changi
8       Choa Chu Kang
9            Clementi
10      Downtown Core
11            Geylang
12            Hougang
13        Jurong East
14        Jurong West
15            Kallang
16             Mandai
17      Marine Parade
18             Newton
19             Novena
20             Outram
21          Pasir Ris
22            Punggol
23         Queenstown
24       River Valley
25             Rochor
26          Sembawang
27           Sengkang
28          Serangoon
29    Singapore River
30           Tampines
31            Tanglin
32          Toa Payoh
33          Woodlands
34             Yishun
35             Others
Name: planning_area, dtype: object

An alternative to the dot-column name way is using the bracket notation.

In [8]:
type_of_dwelling_df['planning_area']

0          Ang Mo Kio
1               Bedok
2              Bishan
3         Bukit Batok
4         Bukit Merah
5       Bukit Panjang
6         Bukit Timah
7              Changi
8       Choa Chu Kang
9            Clementi
10      Downtown Core
11            Geylang
12            Hougang
13        Jurong East
14        Jurong West
15            Kallang
16             Mandai
17      Marine Parade
18             Newton
19             Novena
20             Outram
21          Pasir Ris
22            Punggol
23         Queenstown
24       River Valley
25             Rochor
26          Sembawang
27           Sengkang
28          Serangoon
29    Singapore River
30           Tampines
31            Tanglin
32          Toa Payoh
33          Woodlands
34             Yishun
35             Others
Name: planning_area, dtype: object

The bracket notation can also be used to pull a list of columns.

In [9]:
type_of_dwelling_df[['planning_area','others']]

planning_area  others
0        Ang Mo Kio     391
1             Bedok     611
2            Bishan      53
3       Bukit Batok     108
4       Bukit Merah     344
5     Bukit Panjang       0
6       Bukit Timah     120
7            Changi      41
8     Choa Chu Kang       0
9          Clementi     169
10    Downtown Core       6
11          Geylang     577
12          Hougang     357
13      Jurong East     130
14      Jurong West      75
15          Kallang     547
16           Mandai      18
17    Marine Parade     209
18           Newton       7
19           Novena     166
20           Outram      67
21        Pasir Ris       0
22          Punggol       0
23       Queenstown     133
24     River Valley      13
25           Rochor     218
26        Sembawang      37
27         Sengkang       0
28        Serangoon     130
29  Singapore River       0
30         Tampines     180
31          Tanglin      92
32        Toa Payoh     285
33        Woodlands      77
34           Yishun     137
35           Others      31

#### 3.2 Calculations
We can perform calculations like finding out the number of households in each planning area. 

In [10]:
total = type_of_dwelling_df.flats_1_2_room \
        + type_of_dwelling_df.flats_3_room \
        + type_of_dwelling_df.flats_4_room \
        + type_of_dwelling_df.flats_5_room_and_executive \
        + type_of_dwelling_df.condominiums_and_other_apartments \
        + type_of_dwelling_df.landed_properties \
        + type_of_dwelling_df.others

total

0     59705
1     91225
2     27030
3     44133
4     55626
5     36338
6     20658
7       778
8     47912
9     30792
10     1429
11    39897
12    61539
13    26292
14    78082
15    35474
16      488
17    15498
18     2174
19    15207
20     7368
21    36058
22    18043
23    34364
24     2747
25     5256
26    20612
27    47787
28    35845
29      798
30    73591
31     5587
32    41759
33    68279
34    53576
35     1024
dtype: int64

Instead of writing a code to do a row-wise sum. We can also call the "sum" method to do it. The "axis" of the sum column indicates if it is supposed to perform a row-wise sum (1) or a column-wise sum (0).

In [11]:
numerical_columns = type_of_dwelling_df.columns[1:]
type_of_dwelling_df[numerical_columns].sum(axis=1)

0     59705
1     91225
2     27030
3     44133
4     55626
5     36338
6     20658
7       778
8     47912
9     30792
10     1429
11    39897
12    61539
13    26292
14    78082
15    35474
16      488
17    15498
18     2174
19    15207
20     7368
21    36058
22    18043
23    34364
24     2747
25     5256
26    20612
27    47787
28    35845
29      798
30    73591
31     5587
32    41759
33    68279
34    53576
35     1024
dtype: int64

We can also update each column to be a proportion of the total instead of raw number.

In [12]:
for c in numerical_columns:
    type_of_dwelling_df[c] /=total
type_of_dwelling_df.head()

planning_area  flats_1_2_room  flats_3_room  flats_4_room  \
0    Ang Mo Kio        0.085788      0.436982      0.234369   
1         Bedok        0.046402      0.260806      0.237577   
2        Bishan        0.016426      0.087717      0.367000   
3   Bukit Batok        0.000612      0.248567      0.345433   
4   Bukit Merah        0.203484      0.301532      0.265829   

   flats_5_room_and_executive  condominiums_and_other_apartments  \
0                    0.105033                           0.061318   
1                    0.153500                           0.166051   
2                    0.277839                           0.132075   
3                    0.185168                           0.191399   
4                    0.171395                           0.048700   

   landed_properties    others  
0           0.069961  0.006549  
1           0.128967  0.006698  
2           0.116981  0.001961  
3           0.026375  0.002447  
4           0.002876  0.006184

#### 3.3 Creating new data frame

In [13]:
vif_data = pd.DataFrame()
vif_data

Empty DataFrame
Columns: []
Index: []

#### 3.4 Creating a new column

In [14]:
vif_data["feature"] = numerical_columns
vif_data

feature
0                     flats_1_2_room
1                       flats_3_room
2                       flats_4_room
3         flats_5_room_and_executive
4  condominiums_and_other_apartments
5                  landed_properties
6                             others

In this example, we are using the "variance_inflation_factor" function to calculate the [VIF](https://en.wikipedia.org/wiki/Variance_inflation_factor) (Variance inflation factor)  which is a measurement of multicolliearity introduced by a column. Multicolliearity occurs when one more more columns are providing redundant data. In this example, we obtained the proportion of each type of dwelling. Since they are proportions, their total must add up to 1. Therefore, if one column is removed, no information is lost. As an exmaple, if a + b + c = 1, if a and b are known, we can always calculate c.

Another example of multicolliearity is when two or more columns are highly correlated. Examples of perfect corelation are:
1. Having the two column high in cm and another call height in inches.
2. Having one column qty and the other column cost and the unit cost are all the same.

A rule of thumb is that if a feature has VIF>10 then multicollinearity is high. A cutoff of 5 is also commonly used.

In [15]:
vif_data["VIF"] = [variance_inflation_factor(type_of_dwelling_df[numerical_columns].values, i) for i in range(len(numerical_columns))]        
vif_data

feature        VIF
0                     flats_1_2_room   1.829469
1                       flats_3_room   5.316352
2                       flats_4_room  14.428257
3         flats_5_room_and_executive   8.986825
4  condominiums_and_other_apartments   1.891574
5                  landed_properties   2.144126
6                             others   2.286549

#### 3.5 Removing column(s)

In [16]:
type_of_dwelling_df.drop('flats_4_room', inplace=True, axis=1)

In this below code chunk, we are importing the the "T22" sheet in the "t18-22.xls", making the numbers a proportion of the area's total household,  keeping only the "Planning Area" and "$10,000 & over" columns and rename the two columns appropriately.

In [17]:
household_income_df = pd.read_excel('t18-22.xls', 'T22')
total = household_income_df[household_income_df.columns[1:]].sum(axis=1)
for c in household_income_df.columns:
    if c != 'Planning Area':
        household_income_df[c] = household_income_df[c] / total
        
household_income_df = household_income_df[['Planning Area','$10,000 & Over']]
household_income_df.columns = ['planning_area','over_10000']



In [18]:
#### 3.6 Joining two dataframe

In [19]:
df = type_of_dwelling_df.merge(household_income_df, left_on='planning_area', right_on='planning_area')
df.head()

planning_area  flats_1_2_room  flats_3_room  flats_5_room_and_executive  \
0    Ang Mo Kio        0.085788      0.436982                    0.105033   
1         Bedok        0.046402      0.260806                    0.153500   
2        Bishan        0.016426      0.087717                    0.277839   
3   Bukit Batok        0.000612      0.248567                    0.185168   
4   Bukit Merah        0.203484      0.301532                    0.171395   

   condominiums_and_other_apartments  landed_properties    others  over_10000  
0                           0.061318           0.069961  0.006549    0.203283  
1                           0.166051           0.128967  0.006698    0.277134  
2                           0.132075           0.116981  0.001961    0.373834  
3                           0.191399           0.026375  0.002447    0.296558  
4                           0.048700           0.002876  0.006184    0.192083

## 4. Basic Data analytics
#### 4.1 Scikit learn aka SK-learn
The [scikit learn](https://scikit-learn.org/stable/) package contains modules that is commonly used in machine learning / data analytics. In this example, we are using the "linear_model" module. 

In [20]:
from sklearn import linear_model
from sklearn import metrics

#### 4.2 Preparing data
X and y are commonly used variable to indict the features (X) and the thing we are learning to predict(y).

In [21]:
feature_columns = df.columns[1:-1]
target_column = df.columns[-1]
X = df[feature_columns]
y = df[target_column]

print(feature_columns)
print(target_column)

Index(['flats_1_2_room', 'flats_3_room', 'flats_5_room_and_executive',
       'condominiums_and_other_apartments', 'landed_properties', 'others'],
      dtype='object')
over_10000


#### 4.3 Fitting a Linear Regression model 

In [22]:
reg = linear_model.LinearRegression()
reg = reg.fit(X, y)

#### 4.4 Predicting using fitted model

In [23]:
y_pred = reg.predict(X)
df["prediction"] = y_pred

df

planning_area  flats_1_2_room  flats_3_room  flats_5_room_and_executive  \
0        Ang Mo Kio        0.085788      0.436982                    0.105033   
1             Bedok        0.046402      0.260806                    0.153500   
2            Bishan        0.016426      0.087717                    0.277839   
3       Bukit Batok        0.000612      0.248567                    0.185168   
4       Bukit Merah        0.203484      0.301532                    0.171395   
5     Bukit Panjang        0.000000      0.077302                    0.374842   
6       Bukit Timah        0.000000      0.019992                    0.050634   
7            Changi        0.000000      0.196658                    0.007712   
8     Choa Chu Kang        0.006241      0.027717                    0.413738   
9          Clementi        0.030917      0.409489                    0.106261   
10    Downtown Core        0.000000      0.403779                    0.014696   
11          Geylang        0.108103      0.312555                    0.119207   
12          Hougang        0.005444      0.160289                    0.251678   
13      Jurong East        0.006390      0.278374                    0.315799   
14      Jurong West        0.016508      0.150918                    0.379652   
15          Kallang        0.181260      0.301629                    0.114168   
16           Mandai        0.000000      0.000000                    0.000000   
17    Marine Parade        0.090463      0.202349                    0.116983   
18           Newton        0.000000      0.000000                    0.000000   
19           Novena        0.049056      0.208259                    0.100151   
20           Outram        0.460233      0.312296                    0.009908   
21        Pasir Ris        0.000693      0.000000                    0.483221   
22          Punggol        0.000000      0.000000                    0.669789   
23       Queenstown        0.111192      0.439908                    0.110552   
24     River Valley        0.000000      0.000000                    0.000000   
25           Rochor        0.063737      0.377283                    0.040715   
26        Sembawang        0.000000      0.000000                    0.545702   
27         Sengkang        0.001444      0.005106                    0.526419   
28        Serangoon        0.004910      0.130060                    0.181392   
29  Singapore River        0.000000      0.000000                    0.000000   
30         Tampines        0.005789      0.175579                    0.311954   
31          Tanglin        0.000000      0.004296                    0.012708   
32        Toa Payoh        0.124620      0.392179                    0.158002   
33        Woodlands        0.021749      0.086674                    0.406509   
34           Yishun        0.004909      0.258437                    0.197775   
35           Others        0.000000      0.000000                    0.000000   

    condominiums_and_other_apartments  landed_properties    others  \
0                            0.061318           0.069961  0.006549   
1                            0.166051           0.128967  0.006698   
2                            0.132075           0.116981  0.001961   
3                            0.191399           0.026375  0.002447   
4                            0.048700           0.002876  0.006184   
5                            0.114783           0.020777  0.000000   
6                            0.491964           0.391180  0.005809   
7                            0.616967           0.125964  0.052699   
8                            0.089268           0.013107  0.000000   
9                            0.131430           0.054332  0.005488   
10                           0.379286           0.000000  0.004199   
11                           0.151340           0.053513  0.014462   
12                           0.077479           0.089147  0.005801   
13                          

#### 4.5 Evaulating the model
The most basic evaluation matrics is the mean absolute error.

In [24]:
metrics.mean_absolute_error(y, y_pred)

0.031079116359947423

Other evaluation include RMSE (Root mean square error)

In [25]:
metrics.mean_squared_error(y, y_pred)

0.0017385535285343113

And R-square value. The r-square, measure how much of the variation in the data is explained by the model. A r-square value of 0 means that the model is as good as just predicting the average, and that none of the variation above or below the average is explained by the model.

In [26]:
reg.score(X, y)

0.9141875319454712

#### 4.6 Interpreting the results

In [27]:
reg.intercept_

0.1717522032461692

In [28]:
pd.DataFrame(df.columns[1:-2], reg.coef_)

0
-0.066268                     flats_1_2_room
-0.044635                       flats_3_room
 0.134823         flats_5_room_and_executive
 0.407940  condominiums_and_other_apartments
 0.623853                  landed_properties
-3.499031                             others

The formula generated by the model is as follows:  
Prediction = 0.1717522032461692  
&emsp;&emsp;&emsp;&emsp;&emsp;- 3.499031 * others  
&emsp;&emsp;&emsp;&emsp;&emsp;+ 0.623853 * landed_properties  
&emsp;&emsp;&emsp;&emsp;&emsp;+ 0.407940 * condominiums_and_other_apartments  
&emsp;&emsp;&emsp;&emsp;&emsp;+ 0.134823 * flats_5_room_and_executive  
&emsp;&emsp;&emsp;&emsp;&emsp;- 0.066268 * flats_1_2_room  
&emsp;&emsp;&emsp;&emsp;&emsp;- 0.044635 * flats_3_room  

From this we can say that the model suggests that:
- "Others" has the highest influence on the prediction the greater the value of "others" the smaller the prediction.
- "landed_properties" has the 2nd highest influence. The bigger the value of "landed_propertes" the bigger the prediction.
- "condominiums_and_other_apartments" has the 3rd highest influence. The bigger the value of "condominiums_and_other_apartments" the bigger the prediction.
- For "flats_1_2_room" and "flats_3_room", the larger the value the smaller the prediction.